MimeKit/Cryptography/BouncyCastleSecureMimeContext.cs

```diff
@@ -1225,73 +1225,6 @@ public override Stream Verify (Stream signedData, out DigitalSignatureCollection
 			}
 		}
 
-		class CmsRecipientInfoGenerator : RecipientInfoGenerator
-		{
-			readonly CmsRecipient recipient;
-
-			public CmsRecipientInfoGenerator (CmsRecipient recipient)
-			{
-				this.recipient = recipient;
-			}
-
-			static IWrapper CreateWrapper (AlgorithmIdentifier keyExchangeAlgorithm)
-			{
-				string name;
-
-				if (keyExchangeAlgorithm.Algorithm.Id.Equals (PkcsObjectIdentifiers.IdRsaesOaep.Id, StringComparison.Ordinal)) {
-					var oaepParameters = RsaesOaepParameters.GetInstance (keyExchangeAlgorithm.Parameters);
-					name = "RSA//OAEPWITH" + DigestUtilities.GetAlgorithmName (oaepParameters.HashAlgorithm.Algorithm) + "ANDMGF1Padding";
-				} else if (keyExchangeAlgorithm.Algorithm.Id.Equals (PkcsObjectIdentifiers.RsaEncryption.Id, StringComparison.Ordinal)) {
-					name = "RSA//PKCS1Padding";
-				} else {
-					name = keyExchangeAlgorithm.Algorithm.Id;
-				}
-
-				return WrapperUtilities.GetWrapper (name);
-			}
-
-			static byte[] GenerateWrappedKey (KeyParameter contentEncryptionKey, AlgorithmIdentifier keyEncryptionAlgorithm, AsymmetricKeyParameter publicKey, SecureRandom random)
-			{
-				var keyWrapper = CreateWrapper (keyEncryptionAlgorithm);
-				var keyBytes = contentEncryptionKey.GetKey ();
-
-				keyWrapper.Init (true, new ParametersWithRandom (publicKey, random));
-
-				return keyWrapper.Wrap (keyBytes, 0, keyBytes.Length);
-			}
-
-			public RecipientInfo Generate (KeyParameter contentEncryptionKey, SecureRandom random)
-			{
-				var tbs = Asn1Object.FromByteArray (recipient.Certificate.GetTbsCertificate ());
-				var certificate = TbsCertificateStructure.GetInstance (tbs);
-				var publicKey = recipient.Certificate.GetPublicKey ();
-				var publicKeyInfo = certificate.SubjectPublicKeyInfo;
-				AlgorithmIdentifier keyEncryptionAlgorithm;
-
-				if (publicKey is RsaKeyParameters && recipient.RsaEncryptionPadding?.Scheme == RsaEncryptionPaddingScheme.Oaep) {
```

Comment:
It looks like the BC implementation should handle this for you just fine:

https://github.com/bcgit/bc-csharp/blame/7e3362065e144c4e211eea30f237b106183d70b4/crypto/src/cms/KeyTransRecipientInformation.cs#L67-L74

Reply:
Right, so it uses the certificate to determine PKCS#1v1.5 vs OAEP as opposed to allowing users to override it.

We need to be able to override that.

Reply:
Ahh yes, the algorithm that keeps on giving no end of incompatibilities. I'll need to think about that more.

Reply:
I see... The linked class above is for recipients (decoding this message), not used during generating a message. I don't see an easy way around it, so I'll propose a PR back to bc-csharp.

Reply:
@jstedfast If all goes well, this should show up in `89e04f3a` in bc-csharp and the next release. :-)